sentinel-ai-session/src/main/java/com/phonepe/sentinel/session/AgentSessionExtension.java

```diff
@@ -121,8 +141,55 @@ public void consume(JsonNode output, A agent) {
         }
     }
 
+
+    @Override
+    public List<AgentMessage> messages(R request, AgentRunContext<R> metadata, A agent) {
+        if (isHistoryEnabled()) {
+            final var sessionId = AgentUtils.sessionId(metadata);
+            if (!Strings.isNullOrEmpty(sessionId)) {
+                return historyStore.history(sessionId).map(History::toAgentMessages).orElse(List.of());
+
+            }
+        }
+        return List.of();
+    }
+
     @Override
     public void onExtensionRegistrationCompleted(A agent) {
-        //Nothing to do here for now
+        agent.onRequestCompleted().connect(this::extractAndStoreHistory);
+    }
+
+    private void extractAndStoreHistory(Agent.ProcessingCompletedData<R, T, A> data) {
+        if (isHistoryEnabled()) {
+            final var sessionId = AgentUtils.sessionId(data.getContext());
+            if (!Strings.isNullOrEmpty(sessionId) || data.getOutput().getNewMessages().isEmpty()) {
+                var messages = historyStore.history(sessionId).map(History::getMessages)
+                        .orElse(Lists.newArrayList());
+
+
+                AgentMessage latestPrompt = data.getOutput().getAllMessages().stream()
+                        .filter(agentMessage -> agentMessage.getMessageType()
+                                == AgentMessageType.USER_PROMPT_REQUEST_MESSAGE)
+                        .reduce((first, second) -> second)
+                        .orElse(null);
+                var newMessages = data.getOutput().getNewMessages();
+                if (!Objects.isNull(latestPrompt)) {
+                    newMessages.add(latestPrompt);
+                }
+                messages.add(newMessages);
+                historyStore.saveHistory(History.builder()
+                        .sessionId(sessionId)
+                        .messages(messages)
```

Comment:
Similarly we might want to allow user to pass a function in the setup that can be used to add `metadata` field to history if they want